pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,140 @@
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:   operatorClient,
+		nodeLister:       kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion: semver.MustParse(status.VersionForOperandFromEnv()),
+	}
+	// check if major version updated and no one updated this code
+	if c.apiServerVersion.Minor < 2 {
+		// minor version was 0 or 1, which most likely means the major version has been updated
+		panic(fmt.Errorf("unable to determine kubelet version skew limit"))
+	}
+	c.Controller = factory.New().
+		WithSync(c.sync).
+		WithInformers(kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Informer()).
+		ToController("KubeletVersionSkewController", recorder.WithComponentSuffix("kubelet-version-skew-controller"))
+	return c
+}
+
+type kubeletVersionSkewController struct {
+	factory.Controller
+	operatorClient   v1helpers.OperatorClient
+	nodeLister       corev1listers.NodeLister
+	apiServerVersion semver.Version
+}
+
+func (c *kubeletVersionSkewController) sync(_ context.Context, _ factory.SyncContext) error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(operatorSpec.ManagementState) {
+		return nil
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errorMsgs []string
+	var skewedToTheLimitMsgs []string
+	var skewedButOKMsgs []string
+	var syncedMsgs []string
+	var unsupportedMsgs []string
+
+	// for each node, check kubelet version
+	for _, node := range nodes {
+		kubeletVersion, err := nodeKubeletVersion(node)
+		skew := int(kubeletVersion.Minor - c.apiServerVersion.Minor)
+		switch {
+		case err != nil:
+			errorMsgs = append(errorMsgs, fmt.Sprintf("unable to determine kubelet version on node %s: %s", node.Name, err))
+		case skew <= -2: // behind too far
+			skewedToTheLimitMsgs = append(skewedToTheLimitMsgs, fmt.Sprintf("Kubelet version (%s) on node %s is behind the expected API server version (%s) by at least the maximum allowable skew.", kubeletVersion, node.Name, c.apiServerVersion))
+		case skew == -1: // behind, but ok
+			skewedButOKMsgs = append(skewedButOKMsgs, fmt.Sprintf("Kubelet version (%s) on node %s is behind the expected API server version (%s), but within the maximum allowable skew.", kubeletVersion, node.Name, c.apiServerVersion))
+		case skew == 0: // synced
+			syncedMsgs = append(syncedMsgs, fmt.Sprintf("Kublet version on node %s is in synce with the expected API server version.", node.Name))
+		default: // kubelet version newer than api server version. possibly a rollback.
+			unsupportedMsgs = append(unsupportedMsgs, fmt.Sprintf("Kubelet version (%s) on node %s is ahead of the expected API server version (%s).", kubeletVersion, node.Name, c.apiServerVersion))
+		}
+	}
+
+	// update status
+
+	// combine all msgs
```

Comment:
nit: do we need the above comments - it seems obvious

Reply:
Deleted.